### PR TITLE
improve DFT APIs

### PR DESCRIPTION
**Pull request by wxj6000**
Added functionalities
- Enhance DFT methods with dispersion correction, such as `mf = rks.RKS(mol, xc='wb97m-d3bj')`. 
- Support the customization of DFT methods via specifying the attributes of mean-field objects
  - Disable nlc contributions with `mf.nlc = False`.
  - Specify DFT xc methods and dispersion correction with `mf.xc='b3lyp'` and `mf.disp = 'd3bj'`. 
  - Specify DFT xc method with `mf.xc='b3lyp-d3bj'`.

### CODE REVIEW

File: pyscf/hessian/dispersion.py [hunk 1]
```diff
@@ -23,7 +23,6 @@
 
 import numpy
 from pyscf.dft.rks import KohnShamDFT
-from pyscf.dft import dft_parser
 
 def get_dispersion(hessobj, disp_version=None, with_3body=False):
```
Comment (wxj6000) [inline pyscf/hessian/dispersion.py:27]: with_3body = True

File: pyscf/scf/dispersion.py [hunk 2]
```diff
@@ -21,33 +21,43 @@
 '''
 
 from pyscf.dft.rks import KohnShamDFT
-from pyscf.dft import dft_parser
 
-def get_dispersion(mf, disp_version=None):
+def get_dispersion(mf, disp_version=None, with_3body=False):
     try:
```
Comment (wxj6000) [inline pyscf/scf/dispersion.py:26]: add documents

File: pyscf/scf/dispersion.py [hunk 3]
```diff
@@ -21,33 +21,43 @@
...
-        disp_version = disp
-        # dispersion version can be customized via mf.disp
-        if hasattr(mf, 'disp') and mf.disp is not None:
-            disp_version = mf.disp
-    if disp_version is None:
-        return 0.0
+    # overwrite method if method exists in disp_version
+    if ',' in disp_version:
+        disp_version, method = disp_version.split(',')
+
```
Comment (wxj6000) [inline pyscf/scf/dispersion.py:48]: assert disp_version

File: pyscf/scf/dispersion.py [hunk 4]
```diff
@@ -21,33 +21,43 @@
...
+        else:
+            with_3body = False
 
-    # 3-body contribution can be disabled with mf.disp_with_3body
-    if hasattr(mf, 'disp_with_3body') and mf.disp_with_3body is not None:
-        with_3body = mf.disp_with_3body
+    if hasattr(mf, 'nlc') and mf.nlc not in [False, '']:
+        import warnings
+        warnings.warn('NLC is incompatiable with dispersion correction.')
+        import traceback
```
Comment (wxj6000) [inline pyscf/scf/dispersion.py:59]: delete this

File: pyscf/dft/rks.py [hunk 5]
```diff
@@ -323,10 +323,20 @@ class KohnShamDFT:
...
     def __init__(self, xc='LDA,VWN'):
-        self.xc = xc
-        self.disp = None
-        self.disp_with_3body = None
         self.nlc = ''
+        self.disp = None
+
+        xc, is_nlc, disp = dft_parser.parse_dft(xc)
+        self.xc = xc
+
```
Comment (wxj6000) [inline pyscf/dft/rks.py:331]: simplify the init logic

File: pyscf/dft/rks.py [hunk 6]
```diff
@@ -323,10 +323,20 @@ class KohnShamDFT:
...
+        # overwrite nlc if disabled
+        if is_nlc is False: self.nlc = False
+
+        # overwrite dispersion correction if given
+        method, disp, with_3body = disp
+        if disp: self.disp = disp + ',' + method
+        self.disp_with_3body = with_3body
```
Comment (wxj6000) [inline pyscf/dft/rks.py:338]: remove disp_with_3body

File: pyscf/hessian/dispersion.py [hunk 7]
```diff
@@ -33,33 +32,41 @@ def get_dispersion(hessobj, disp_version=None, with_3body=False):
...
-        with_3body = mf.disp_with_3body
+    if isinstance(mf, KohnShamDFT):
+        method = mf.xc
+    else:
+        method = 'hf'
+
+    # overwrite method if method exists in disp_version
+    if ',' in disp_version:
+        disp_version, method = disp_version.split(',')
+
```
Comment (wxj6000) [inline pyscf/hessian/dispersion.py:53]: parse 2b and ATM

File: pyscf/tdscf/uhf.py [hunk 8]
```diff
@@ -181,7 +181,7 @@ def add_hf_(a, b, hyb=1):
         from pyscf.dft import xc_deriv
         ni = mf._numint
         ni.libxc.test_deriv_order(mf.xc, 2, raise_error=True)
-        if mf.nlc or ni.libxc.is_nlc(mf.xc):
```
Comment (wxj6000) [inline pyscf/tdscf/uhf.py:184]: is not -> ! =

Comment (wxj6000) [reply pyscf/tdscf/uhf.py:184]: define a new function `do_nlc` and explain

File: pyscf/dft/rks.py [hunk 9]
```diff
@@ -361,7 +371,7 @@ def dump_flags(self, verbose=None):
 
         self.grids.dump_flags(verbose)
 
-        if self.nlc or self._numint.libxc.is_nlc(self.xc):
+        if self.nlc is not False and (self.nlc or self._numint.libxc.is_nlc(self.xc)):
             log.info('** Following is NLC and NLC Grids **')
             if self.nlc:
                 log.info('NLC functional = %s', self.nlc)
```
Comment (wxj6000) [inline pyscf/dft/rks.py:377]: add warning in libxc.py if with dispersion

File: examples/dft/16-dft_d3.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+# d3zero), and which XC type of dispersion to target at (like b3lyp, wb97, hf)
+# The two parameters can be configured in the disp attribute, separated by ","
+mf.disp = 'd3,b3lyp'
+mf.disp = 'd4,wb97m'
+mf.disp = 'd3bj,hf'
+
+# If the xc-type is not specified in the disp, the DFT code will automatically
+# employ the .xc attribute as the xc-type parameter.
+mf.xc = 'wb97x'
+mf.disp = 'd3' # == 'd3,wb97x'
```
Comment (hebrewsnabla) [inline examples/dft/16-dft_d3.py:66]: Is this input desired to be a correct example of wB97X-D3? The SCF part of wB97X-D3 is not wB97X.

Comment (hebrewsnabla) [reply examples/dft/16-dft_d3.py:66]: I don't understand why this is resolved since there's no changes to those examples.

Comment (susilehtola) [reply examples/dft/16-dft_d3.py:66]: I concur, this is dangerous

File: pyscf/dft/test/test_libxc.py [hunk 11]
```diff
@@ -343,18 +343,10 @@ def test_m06(self):
...
-
-        self.assertEqual(parse_dft('wb97m-d3bjatm'), ('wb97m-v', False, ('wb97m', 'd3bj', True)))
-        self.assertTrue(not dft.libxc.is_nlc('wb97m-d3bjatm'))
-
-        self.assertEqual(parse_dft('wb97x-d3zero2b'), ('wb97x', None, ('wb97x', 'd3zero', False)))
-        self.assertTrue(not dft.libxc.is_nlc('wb97x-d3zero2b'))
+        self.assertEqual(parse_dft('wb97m-d3bj'), ('wb97m-v', False, 'd3bj,wb97m'))
+        self.assertEqual(parse_dft('b3lyp-d3zerom'), ('b3lyp', '', 'd3zerom,b3lyp'))
+        self.assertEqual(parse_dft('wb97x-d3bj'), ('wb97x-v', False, 'd3bj,wb97x'))
+        self.assertEqual(parse_dft('wb97x-d3zero2b'), ('wb97x', '', 'd3zero2b,wb97x'))
```
Comment (susilehtola) [inline pyscf/dft/test/test_libxc.py:349]: same here

**Comment by sunqm** [comment]
@hebrewsnabla @susilehtola Hope the updates look reasonable now.

**Comment by hebrewsnabla** [comment]
Looks good to me.

**Comment by susilehtola** [comment]
I think otherwise it looked reasonable

**Review by sunqm — APPROVED** [review]

**Comment by sunqm** [comment]
This PR will change the initialization of DFT classes. The two DFT initialization methods
```
mf = dft.KS(mol, xc='xxxx')
```
and
```
mf = dft.KS(mol)
mf.xc = 'xxxx'
```
will be no longer equivalent for all XC functionals. When dft is set up with dispersion correction from d3/d4, only the former method works. @fishjojo @tberkel Please take a look if this design proposed in this PR looks reasonable to you.
I'll wait for one more week for any comments or suggestions before merging this PR.

File: examples/dft/16-dft_d3.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+mf.disp = 'd3,b3lyp'
+mf.disp = 'd4,wb97m'
+mf.disp = 'd3bj,hf'
+
+# If the xc parameter is not specified in the disp, the DFT code will automatically
+# employ the .xc attribute as the xc-type parameter.
+mf.xc = 'b3lyp'
+mf.disp = 'd3bj' # == 'd3bj,b3lyp'
+
+# Please note that the second parameter of mf.disp can be different to mf.xc.
```
Comment (fishjojo) [inline examples/dft/16-dft_d3.py:87]: maybe print a warning when this happens, since this may be due to mistake input.

**Review by fishjojo — APPROVED** [review]

File: examples/dft/16-dft_d3.py (hunk 12 above)
Comment (wxj6000) [reply examples/dft/16-dft_d3.py:87]: The xc parser will raise error if mf.xc contains 'd3' or 'd4'. https://github.com/wxj6000/pyscf/blob/enhance_dft_api/pyscf/dft/libxc.py#L1101

Comment (fishjojo) [reply examples/dft/16-dft_d3.py:87]: I mean when mf.xc != xc in mf.disp

Comment (wxj6000) [reply examples/dft/16-dft_d3.py:87]: I don't know. :smile: I am not very confident enough to provide the proper warning message for all the cases. mf.xc may be different from xc in mf.disp as desired. For instance, wb97m-d3bj. The definition of xc can change in different packages. For instance B3LYP.

But I added some prints in dispersion modules.

**Comment by tberkel** [comment]
> This PR will change the initialization of DFT classes. The two DFT initialization methods
> 
> ```
> mf = dft.KS(mol, xc='xxxx')
> ```
> 
> and
> 
> ```
> mf = dft.KS(mol)
> mf.xc = 'xxxx'
> ```
> 
> will be no longer equivalent for all XC functionals. When dft is set up with dispersion correction from d3/d4, only the former method works. @fishjojo @tberkel Please take a look if this design proposed in this PR looks reasonable to you. I'll wait for one more week for any comments or suggestions before merging this PR.

I admit I haven't followed the discussion closely and/or looked deep into the PR code, but this inequivalence is worrisome and I guess will cause occasional confusion. It's probably the only place in PySCF where passing a kwarg and setting an attribute for the same variable can have different behavior.

**Comment by wxj6000** [comment]
> about the inconsistency issue raised above, can we use the `@setter` mechanism in python such that
> 
> ```
> mf = dft.KS(mol)
> mf.xc = 'xxxx'
> ```
> 
> calls the setter of xc to make it equivalent to
> 
> ```
> mf = dft.KS(mol, xc='xxxx')
> ```
> 
> ?
> 
> If not possible, I would suggest simply ban
> 
> ```
> mf = dft.KS(mol)
> mf.xc = 'xxxx'
> ```
> 
> and allow only
> 
> ```
> mf = dft.KS(mol, xc='xxxx')
> ```

@wenyan4work We are drafting a simplified version. It should be ready this week.

File: pyscf/dft/rks.py [hunk 13]
```diff
@@ -374,6 +374,28 @@ def dump_flags(self, verbose=None):
...
+            if self.nlc == '' or self.nlc == 0:
+                return False
+            else:
+                raise RuntimeError('Conflict found between dispersion and xc.')
+
+        # If nlc is disabled via self.nlc
+        if self.nlc == 0:
+            return False
+
+        xc_has_nlc = self._numint.libxc.is_nlc(xc)
```
Comment (sunqm) [inline pyscf/dft/rks.py:396]: `if self.nl == 'vv10'`

File: examples/dft/16-dft_d3.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+mf.xc = 'wb97x-d3bj'
+mf.kernel()
+
+# 2. mf.xc, mf.nlc, mf.disp = 'xc-keyword', 0, 'd3'
+#   Manually control the dispersion calculations, where nlc is muted, and disp
+#   is set to a particular version of DFT-D3.
+mf = mol.KS()
+mf.xc = 'wb97x-v'
+mf.nlc = 0  # this will disable NLC correction.
+mf.disp = 'd3'
```
Comment (wxj6000) [inline examples/dft/16-dft_d3.py:102]: 'd3' is illegal.

File: pyscf/scf/dispersion.py [hunk 15]
```diff
@@ -20,37 +20,138 @@
 dispersion correction for HF and DFT
 '''
 
-from pyscf.dft.rks import KohnShamDFT
-from pyscf.dft import dft_parser
+import warnings
+from functools import lru_cache
+from pyscf.lib import logger
+
+# supported dispersion corrections
```
Comment (wxj6000) [inline pyscf/scf/dispersion.py:27]: add link to dftd3

**Comment by jfreeze95** [comment]
There's an issue with this merged update, it expects a pyscf.dispersion update that isn't in the current main branch this has been merged into. 

Here's an example:
```
from pyscf import gto
from pyscf.hessian import thermo

# First compute nuclear Hessian.
mol = gto.M(
    atom = '''C           1.00516        -0.00613         0.12498
C           2.52764        -0.00573         0.14659
C           3.06976        -0.36885         1.54129
O           3.77997        -1.33992         1.77093
O           2.72723         0.51104         2.51062
N           3.01707        -0.97290        -0.85939
H           0.59853         0.75568         0.79892
H           0.59903        -0.97280         0.44484
H           0.62491         0.21061        -0.87914
H           2.91615         0.98244        -0.11975
H           2.19226         1.26576         2.20107
H           2.72345        -1.91043        -0.56917
H           4.03776        -1.01670        -0.78276
''',
    basis = '631g',
    symmetry = True)
#mf = mol.RHF().run()
mf = mol.KS(xc = 'wb97x-d3bj')
mf.kernel()
```

And here's the error:
```
File ".../miniconda3/lib/python3.11/site-packages/pyscf/scf/dispersion.py", line 140, in get_dispersion
    from pyscf.dispersion import dftd3, dftd4
ModuleNotFoundError: No module named 'pyscf.dispersion'
```

This I have identified as the pip install for pyscf-dispersion needs updated to the latest version of pyscf-dispersion since it currently pulls an older version.

Using the newer version or just changing  line 21 of dftd3.py to "libdftd3 = lib.load_library('libs-dftd3')" gives a new OS related error

`OSError: dlopen(.../miniconda3/lib/python3.11/site-packages/pyscf/lib/libs-dftd3.dylib, 0x0006): tried: '.../miniconda3/lib/python3.11/site-packages/pyscf/lib/libs-dftd3.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/.../miniconda3/lib/python3.11/site-packages/pyscf/lib/libs-dftd3.dylib' (no such file), '.../miniconda3/lib/python3.11/site-packages/pyscf/lib/libs-dftd3.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64'))`

I tried uninstalling and reinstalling with:
`arch -arm64 python3 -m pip install pyscf-dispersion --no-cache`

which presented no change.

I am posting this here and not as an issue because as this was released to the stable release, I think it may be causing many of the dispersion issues I am seeing in the Issue section from the past week.

**Pull request merged**